internal/sync2/throttle_test.go

```diff
@@ -0,0 +1,94 @@
+package sync2
+
+import (
+	"fmt"
+	"io"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+func ExampleThrottle() {
+	throttle := NewThrottle()
+	var wg sync.WaitGroup
+
+	// consumer
+	go func() {
+		defer wg.Done()
+		totalConsumed := int64(0)
+		for {
+			available, err := throttle.ConsumeOrWait(8)
+			if err != nil {
+				return
+			}
+			fmt.Println("- consuming ", available, " total=", totalConsumed)
+			totalConsumed += available
+
+			// do work for available amount
+			time.Sleep(time.Duration(available) * time.Millisecond)
+		}
+	}()
+
+	// producer
+	go func() {
+		defer wg.Done()
+
+		step := int64(8)
+		for total := int64(64); total >= 0; total -= step {
+			err := throttle.ProduceAndWaitUntilBelow(step, step*3)
+			if err != nil {
+				return
+			}
+
+			fmt.Println("+ producing", step, " left=", total)
+			time.Sleep(time.Duration(rand.Intn(8)) * time.Millisecond)
+		}
+
+		throttle.Fail(io.EOF)
+	}()
+
+	wg.Wait()
+
+	fmt.Println("done", throttle.Err())
+}
+
+func TestThrottleBasic(t *testing.T) {
+	throttle := NewThrottle()
+	var stage int64
+	c := make(chan error)
```

Comment:
add a capacity here

Reply:
What do you mean by capacity?

Reply:
`c := make(chan error, 1)` - this ensures that when both goroutine and the main fail, it won't get stuck.